create large movie review reader

cz9779 opened the pull request:
This PR fixes [#225 ](https://github.com/asyml/forte/issues/225). 

### Description of changes
This PR creates a reader for the [large movie review dataset](https://ai.stanford.edu/~amaas/data/sentiment/).

### Possible influences of this PR.
Read sentences and documents from the movie review texts. Add a Document-level sentiment attribute to the base-ontology.

### Test Conducted
Create a unit test for the reader with samples from the large movie review dataset.

hunterhector commented on forte/data/readers/largemovie_reader.py:
```diff
@@ -0,0 +1,116 @@
+        data_pack.pack_name = doc_id
+        data_pack.set_text(doc_text)
+
+        # add documents.
```
A small style suggestion, use full sentences, and capitalize the first character.

hunterhector commented on forte/data/readers/largemovie_reader.py:
```diff
@@ -0,0 +1,116 @@
+
+        pos_dir = os.path.basename(os.path.dirname(file_path))
+        movie_file = os.path.basename(file_path)
+        docid, score = movie_file.split('_')
```
Let's annotate a few more data types here.

And after annotating the type, you will find that re-assigning different data type to the same variable `score` is not suggested.

hunterhector commented on forte/data/readers/largemovie_reader.py:
```diff
@@ -0,0 +1,116 @@
+
+        yield data_pack
+
+    def _cache_key_function(self, data_pack: DataPack) -> str:
```
No, `cache_key_function` take the output of `_collect` to do caching to help construct `data_pack` faster. If you already have a data_pack then you don't need to query the cache.

cz9779 replied:
I see what you mean but I referred it from the `_cache_key_function` function defined in `conllu_ud_reader.py`, I guess it needs to be changed too?

hunterhector replied:
I see, thanks for spotting this. Could you help change that too?

cz9779 replied:
> I see, thanks for spotting this. Could you help change that too?

@hunterhector As I trying to fix it for other readers, I spot that there are more files that used `data_pack` as input, including `ms_macro_passage_reader` and `sst2_reader` that was just merged yesterday by my teammate. Do you think it's better that we create another PR to solve this issue?

hunterhector commented on forte/data/readers/largemovie_reader.py:
```diff
@@ -0,0 +1,116 @@
+
+"""
+A reader to read reviews and IMDb scores from `Large Movie Reviews` dataset for sentiment classification.
+
```
If I download the dataset from the link, how can I use this reader? We can write the command in the docstring to help the users.

hunterhector replied:
I mean you can make it clear how we can use the reader, like what input parameter should be used.

hunterhector commented on forte/data/readers/largemovie_reader.py:
```diff
@@ -0,0 +1,116 @@
+# Copyright 2019 The Forte Authors. All Rights Reserved.
```
I just realize we should change the copyright to 2020 now

cz9779 replied:
> I just realize we should change the copyright to 2020 now

@hunterhector Should I apply the change of 2020 to all the other files in this project?

hunterhector replied:
Let's just use 2020 for new files. I'm currently not sure what I should do about the rest, yet.

hunterhector reviewed (changes requested):
Overall pretty good except the misuse on `_cache_key_function` and a couple of small comments. And for sure you have to make sure the travis is passing.

To test whether `__cache_key_functon` is working, you can try test it with the caching function on.

hunterhector commented on forte/data/readers/largemovie_reader.py:
```diff
@@ -86,31 +94,31 @@ def _parse_pack(self, file_path: str) -> Iterator[DataPack]:
-            raise ValueError("Data pack does not have a document id.")
-        return data_pack.pack_name
+    def _cache_key_function(self, movie_file: str) -> str:
+        # pos0_9.txt, neg3_4.txt
```
This comment would be confusing for other developers.

hunterhector commented:
I see. In this case, It would bit a bit more involved, and actually this
means there is a design/documentation issue. Could you help open a new
issue? We can plan for the PR later. The issue should contain the following
points:

1. Revisit the design of cache_key_function, currently, it is a bit
confusing to implement and the documentation for this is not enough.
2. We should design a general reader test that can be reused for all reader
classes, which will test the caching ability if caching is implemented.



On Wed, Sep 9, 2020 at 5:37 PM cz9779 <notifications@github.com> wrote:

> *@cz9779* commented on this pull request.
> ------------------------------
>
> In forte/data/readers/largemovie_reader.py
> <https://github.com/asyml/forte/pull/237#discussion_r485935987>:
>
> > +        movie_file = os.path.basename(file_path)
> +        docid, score = movie_file.split('_')
> +        doc_id = pos_dir + docid
> +        score = score.split('.')[0]
> +        score = float(score) / 10.0
> +
> +        data_pack.pack_name = doc_id
> +        data_pack.set_text(doc_text)
> +
> +        # add documents.
> +        document:Document = Document(data_pack, 0, len(doc_text))
> +        document.sentiment = {doc_id: score}
> +
> +        yield data_pack
> +
> +    def _cache_key_function(self, data_pack: DataPack) -> str:
>
> I see, thanks for spotting this. Could you help change that too?
>
> @hunterhector <https://github.com/hunterhector> As I tried to fix it for
> other readers, I spot that there are more files that used data_pack as
> input, including ms_macro_passage_reader and sst2_reader that was just
> merged yesterday by my teammate. Do you think it's better that we create
> another PR to solve this issue?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/asyml/forte/pull/237#discussion_r485935987>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHYBN2HR3WNVBVL6D3CSV3SE7YRFANCNFSM4RCVK54A>
> .
>


-- 
Hector, Liu Zhengzhong

Website: https://hunterhector.github.io/
<http://www.linkedin.com/pub/hector-zhengzhong-liu/23/115/406>
LinkedIn: http://www.linkedin.com/pub/hector-zhengzhong-liu/23/115/406

hunterhector reviewed (changes requested):
Just a few small comments.

cz9779 commented:
> I see. In this case, It would bit a bit more involved, and actually this means there is a design/documentation issue. Could you help open a new issue? We can plan for the PR later. The issue should contain the following points: 1. Revisit the design of cache_key_function, currently, it is a bit confusing to implement and the documentation for this is not enough. 2. We should design a general reader test that can be reused for all reader classes, which will test the caching ability if caching is implemented.
> […](#)
> On Wed, Sep 9, 2020 at 5:37 PM cz9779 ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In forte/data/readers/largemovie_reader.py <[#237 (comment)](https://github.com/asyml/forte/pull/237#discussion_r485935987)>: > + movie_file = os.path.basename(file_path) + docid, score = movie_file.split('_') + doc_id = pos_dir + docid + score = score.split('.')[0] + score = float(score) / 10.0 + + data_pack.pack_name = doc_id + data_pack.set_text(doc_text) + + # add documents. + document:Document = Document(data_pack, 0, len(doc_text)) + document.sentiment = {doc_id: score} + + yield data_pack + + def _cache_key_function(self, data_pack: DataPack) -> str: I see, thanks for spotting this. Could you help change that too? @hunterhector <https://github.com/hunterhector> As I tried to fix it for other readers, I spot that there are more files that used data_pack as input, including ms_macro_passage_reader and sst2_reader that was just merged yesterday by my teammate. Do you think it's better that we create another PR to solve this issue? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#237 (comment)](https://github.com/asyml/forte/pull/237#discussion_r485935987)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAHYBN2HR3WNVBVL6D3CSV3SE7YRFANCNFSM4RCVK54A> .
> -- Hector, Liu Zhengzhong Website: https://hunterhector.github.io/ <http://www.linkedin.com/pub/hector-zhengzhong-liu/23/115/406> LinkedIn: http://www.linkedin.com/pub/hector-zhengzhong-liu/23/115/406

I have created a new issue #239(https://github.com/asyml/forte/issues/239#issue-697193132)

cz9779 commented:
I have created a new issue #239 (https://github.com/asyml/forte/issues/239) that links to the cache function in readers.

hunterhector reviewed (approved):
LGTM
